dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/StatsTaskListGenerator.java

```diff
@@ -65,47 +65,40 @@ class StatsTaskListGenerator {
           "table-types-dtl",
           "used-space-details");
 
-  private static final ImmutableList<String> STATSPACK_NAMES =
-      ImmutableList.of("hist-cmd-types-statspack", "sql-stats-statspack");
-
   @Nonnull
   ImmutableList<Task<?>> createTasks(ConnectorArguments arguments, Duration queriedDuration) {
     ImmutableList.Builder<Task<?>> builder = ImmutableList.builder();
     builder.add(new DumpMetadataTask(arguments, scope.formatName()));
     builder.add(new FormatTask(scope.formatName()));
-    List<StatsJdbcTask> jdbcTasks = createJdbcTasks(queriedDuration);
-    builder.addAll(jdbcTasks);
 
-    QueryGroup awrAndCdb = QueryGroup.create(/* required= */ false, AWR, MULTI_TENANT);
-    QueryGroup awrAndNotCdb = QueryGroup.create(/* required= */ false, AWR, SINGLE_TENANT);
-    builder.add(ResultMessageTask.create(awrAndCdb, jdbcTasks));
-    builder.add(ResultMessageTask.create(awrAndNotCdb, jdbcTasks));
+    ArrayList<Task<?>> messageTasks = new ArrayList<>();
+    ImmutableList<OracleSqlList> listsInOrder = ImmutableList.of(awrCdb(), awrDba(), statspack());
 
-    QueryGroup statspackAndCdb = QueryGroup.create(/* required= */ false, STATSPACK, MULTI_TENANT);
-    QueryGroup statspackAndNotCdb =
-        QueryGroup.create(/* required= */ false, STATSPACK, SINGLE_TENANT);
-    builder.add(ResultMessageTask.create(statspackAndCdb, jdbcTasks));
-    builder.add(ResultMessageTask.create(statspackAndNotCdb, jdbcTasks));
+    for (OracleSqlList sqls : listsInOrder) {
+      ImmutableList<Task<?>> messageTasksSoFar = copyOf(messageTasks);
+      ImmutableList<Task<?>> queryTasks =
+          sqls.toTasksIfAllSkipped(queriedDuration, messageTasksSoFar);
+      Task<?> success = ResultMessageTask.createNoFilter(sqls.group(), queryTasks);
+      builder.addAll(queryTasks);
+      messageTasks.add(success);
+    }
+
+    // Summary messages
+    builder.addAll(messageTasks);
```

Comment:
I assume that the order of tasks in the builder is:
queryTasks1, queryTasks2, queryTasks3, messageTasks, nativeTasks

IIUC the execution of the queryTasks2 depends on the first messageTask. How can queryTasks2 check the condition of messageTask, if it is scheduled to be executed later than queryTasks2?

Reply:
True, this is a bug. I forgot about the need to manually order the tasks when I added line 88.

The correct approach is to never add from `messageTasks` to `builder`. Instead, the `success` task should be added while inside the loop.

Reply:
Done